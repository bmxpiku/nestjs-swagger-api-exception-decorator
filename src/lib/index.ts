import { HttpException } from '@nestjs/common';

import { ApiException } from './decorators/api-exception.decorator';
import {
  ExceptionArguments,
  ExceptionOrExceptionArray,
  ExceptionOrExceptionArrayFunc,
} from './interfaces/api-exception.interface';
import { Options, Template } from './interfaces/options.interface';
import { areExceptionsPassedWithoutArrowFunction } from './utils/exception.util';

/**
 * Build your own custom decorator. This enables you to re-use the same template again without the need to specify it
 * over and over again
 *
 * @param template Any object describing the template which should be shown as example value
 * @param globalOptions Specify the content type
 */
export const buildTemplatedApiExceptionDecorator = <T = Template>(
  template: T,
  globalOptions?: Omit<Options<T>, 'template'>,
) => {
  return <Exception extends HttpException>(exceptions: ExceptionArguments<Exception>, options?: Options<T>) => {
    const mergedOptions = ({ ...globalOptions, template, ...options } as unknown) as Options<Template>;

    if (areExceptionsPassedWithoutArrowFunction(exceptions)) {
      return ApiException(exceptions as ExceptionOrExceptionArray<Exception>, mergedOptions);
    } else {
      return ApiException(exceptions as ExceptionOrExceptionArrayFunc<Exception>, mergedOptions);
    }
  };
};

export { ApiException };
