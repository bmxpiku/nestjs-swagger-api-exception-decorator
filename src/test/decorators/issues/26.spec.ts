import { InternalServerErrorException } from '@nestjs/common';
import { DECORATORS } from '@nestjs/swagger/dist/constants';

import { ApiException } from '../../../lib';

class EmailResponseFailedException extends InternalServerErrorException {
  constructor() {
    super('Email Response Failed');
  }
}

class InternalServerException extends InternalServerErrorException {
  constructor() {
    super('Something went wrong');
  }
}

describe('Issue #26', () => {
  describe('given InternalServerError exception and custom exception sharing the same status code', () => {
    it('should attach the exceptions properly without missing another', () => {
      class InternalServerError {
        @ApiException(() => [EmailResponseFailedException, InternalServerException])
        test() {
          return;
        }
      }

      const descriptor = Object.getOwnPropertyDescriptor(InternalServerError.prototype, 'test');
      const meta = Reflect.getMetadata(DECORATORS.API_RESPONSE, descriptor.value);

      expect(meta).toMatchSnapshot();
    });
  });
});
