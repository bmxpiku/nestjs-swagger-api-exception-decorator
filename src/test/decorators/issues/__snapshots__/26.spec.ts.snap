// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Issue #26 given InternalServerError exception and custom exception sharing the same status code should attach the exceptions properly without missing another 1`] = `
Object {
  "500": Object {
    "content": Object {
      "application/json": Object {
        "examples": Object {
          "EmailResponseFailedException": Object {
            "description": "Email Response Failed",
            "value": Object {
              "error": "Internal Server Error",
              "message": "Email Response Failed",
              "statusCode": 500,
            },
          },
          "InternalServerException": Object {
            "description": "Something went wrong",
            "value": Object {
              "error": "Internal Server Error",
              "message": "Something went wrong",
              "statusCode": 500,
            },
          },
        },
        "schema": Object {
          "description": undefined,
          "properties": Object {
            "error": Object {
              "example": "Internal Server Error",
              "type": "string",
            },
            "message": Object {
              "example": "Email Response Failed",
              "type": "string",
            },
            "statusCode": Object {
              "example": 500,
              "type": "number",
            },
          },
          "required": Array [
            "statusCode",
            "message",
          ],
          "type": "object",
        },
      },
    },
    "description": "",
    "isArray": undefined,
    "type": undefined,
  },
}
`;
